File: API/3152_Macd_1_Min_Scalper/CS/Macd1MinScalperStrategy.cs
namespace StockSharp.Samples.Strategies;

using System;

using Ecng.Common;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.Messages;

using StockSharp.Algo.Candles;

/// <summary>
/// Simplified MACD 1 Min Scalper - uses WMA crossover + MACD + Momentum confirmation.
/// </summary>
public class Macd1MinScalperStrategy : Strategy
{
	private WeightedMovingAverage _fastMa;
	private WeightedMovingAverage _slowMa;
	private MovingAverageConvergenceDivergence _macd;
	private Momentum _momentum;

	private decimal _prevFast;
	private decimal _prevSlow;
	private bool _hasPrev;
	private decimal? _entryPrice;

	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		_fastMa = new WeightedMovingAverage { Length = 3 };
		_slowMa = new WeightedMovingAverage { Length = 5 };
		_macd = new MovingAverageConvergenceDivergence();
		_momentum = new Momentum { Length = 14 };

		var subscription = SubscribeCandles(TimeSpan.FromMinutes(5).TimeFrame());
		subscription
			.Bind(ProcessCandle)
			.Start();
	}

	private void ProcessCandle(ICandleMessage candle)
	{
		if (candle.State != CandleStates.Finished)
			return;

		var price = candle.ClosePrice;
		var openTime = candle.OpenTime;

		var fastResult = _fastMa.Process(new DecimalIndicatorValue(_fastMa, price, openTime) { IsFinal = true });
		var slowResult = _slowMa.Process(new DecimalIndicatorValue(_slowMa, price, openTime) { IsFinal = true });
		var macdResult = _macd.Process(new DecimalIndicatorValue(_macd, price, openTime) { IsFinal = true });
		var momResult = _momentum.Process(new DecimalIndicatorValue(_momentum, price, openTime) { IsFinal = true });

		if (!_fastMa.IsFormed || !_slowMa.IsFormed || !_macd.IsFormed || !_momentum.IsFormed)
			return;

		if (fastResult.IsEmpty || slowResult.IsEmpty || macdResult.IsEmpty || momResult.IsEmpty)
			return;

		var fast = fastResult.ToDecimal();
		var slow = slowResult.ToDecimal();
		var macdVal = macdResult.ToDecimal();
		var momVal = momResult.ToDecimal();

		// Risk management
		if (_entryPrice.HasValue && Position != 0)
		{
			var step = Security?.PriceStep ?? 1m;
			if (step <= 0) step = 1m;
			var stopDist = 100 * step;
			var takeDist = 100 * step;

			if (Position > 0 && (candle.ClosePrice <= _entryPrice.Value - stopDist || candle.ClosePrice >= _entryPrice.Value + takeDist))
			{
				SellMarket();
				_entryPrice = null;
				_hasPrev = false;
				_prevFast = fast; _prevSlow = slow;
				return;
			}
			if (Position < 0 && (candle.ClosePrice >= _entryPrice.Value + stopDist || candle.ClosePrice <= _entryPrice.Value - takeDist))
			{
				BuyMarket();
				_entryPrice = null;
				_hasPrev = false;
				_prevFast = fast; _prevSlow = slow;
				return;
			}
		}

		if (_hasPrev)
		{
			// Buy: fast crosses above slow + MACD positive + momentum above 100
			var buySignal = fast > slow && _prevFast <= _prevSlow && macdVal > 0 && momVal > 100;
			// Sell: fast crosses below slow + MACD negative + momentum below 100
			var sellSignal = fast < slow && _prevFast >= _prevSlow && macdVal < 0 && momVal < 100;

			if (buySignal && Position <= 0)
			{
				if (Position < 0)
				{
					BuyMarket();
					_entryPrice = null;
				}
				BuyMarket();
				_entryPrice = candle.ClosePrice;
			}
			else if (sellSignal && Position >= 0)
			{
				if (Position > 0)
				{
					SellMarket();
					_entryPrice = null;
				}
				SellMarket();
				_entryPrice = candle.ClosePrice;
			}
		}

		_prevFast = fast;
		_prevSlow = slow;
		_hasPrev = true;
	}
}

File: API/3101-3200/3152_Macd_1_Min_Scalper/PY/macd_1_min_scalper_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import ExponentialMovingAverage
from StockSharp.Algo.Strategies import Strategy

class macd_1_min_scalper_strategy(Strategy):
    def __init__(self):
        super(macd_1_min_scalper_strategy, self).__init__()
        self._fast_period = self.Param("FastPeriod", 12).SetGreaterThanZero().SetDisplay("Fast Period", "Fast EMA period", "Indicator")
        self._slow_period = self.Param("SlowPeriod", 50).SetGreaterThanZero().SetDisplay("Slow Period", "Slow EMA period", "Indicator")
        self._stop_loss_points = self.Param("StopLossPoints", 200).SetNotNegative().SetDisplay("Stop Loss", "Stop-loss in price steps", "Risk")
        self._take_profit_points = self.Param("TakeProfitPoints", 400).SetNotNegative().SetDisplay("Take Profit", "Take-profit in price steps", "Risk")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(5))).SetDisplay("Candle Type", "Candles", "General")

        self._prev_fast = 0.0
        self._prev_slow = 0.0
        self._entry_price = 0.0
        self._cooldown = 0

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(macd_1_min_scalper_strategy, self).OnReseted()
        self._prev_fast = 0.0
        self._prev_slow = 0.0
        self._entry_price = 0.0
        self._cooldown = 0

    def OnStarted2(self, time):
        super(macd_1_min_scalper_strategy, self).OnStarted2(time)
        self._fast_ind = ExponentialMovingAverage()
        self._fast_ind.Length = self._fast_period.Value
        self._slow_ind = ExponentialMovingAverage()
        self._slow_ind.Length = self._slow_period.Value
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(self._fast_ind, self._slow_ind, self._process_candle).Start()

    def _process_candle(self, candle, fast_val, slow_val):
        if candle.State != CandleStates.Finished:
            return
        fast = float(fast_val)
        slow = float(slow_val)
        if not self._fast_ind.IsFormed or not self._slow_ind.IsFormed:
            self._prev_fast = fast
            self._prev_slow = slow
            return
        if self._cooldown > 0:
            self._cooldown -= 1
            self._prev_fast = fast
            self._prev_slow = slow
            return
        close = float(candle.ClosePrice)
        step = float(self.Security.PriceStep) if self.Security is not None and self.Security.PriceStep is not None else 1.0

        if self.Position > 0 and self._entry_price > 0:
            if self._stop_loss_points.Value > 0 and close <= self._entry_price - self._stop_loss_points.Value * step:
                self.SellMarket()
                self._entry_price = 0.0
                self._cooldown = 100
                self._prev_fast = fast
                self._prev_slow = slow
                return
            if self._take_profit_points.Value > 0 and close >= self._entry_price + self._take_profit_points.Value * step:
                self.SellMarket()
                self._entry_price = 0.0
                self._cooldown = 100
                self._prev_fast = fast
                self._prev_slow = slow
                return
        elif self.Position < 0 and self._entry_price > 0:
            if self._stop_loss_points.Value > 0 and close >= self._entry_price + self._stop_loss_points.Value * step:
                self.BuyMarket()
                self._entry_price = 0.0
                self._cooldown = 100
                self._prev_fast = fast
                self._prev_slow = slow
                return
            if self._take_profit_points.Value > 0 and close <= self._entry_price - self._take_profit_points.Value * step:
                self.BuyMarket()
                self._entry_price = 0.0
                self._cooldown = 100
                self._prev_fast = fast
                self._prev_slow = slow
                return

        if self._prev_fast <= self._prev_slow and fast > slow and self.Position <= 0:
            if self.Position < 0:
                self.BuyMarket()
            self.BuyMarket()
            self._entry_price = close
            self._cooldown = 100
        elif self._prev_fast >= self._prev_slow and fast < slow and self.Position >= 0:
            if self.Position > 0:
                self.SellMarket()
            self.SellMarket()
            self._entry_price = close
            self._cooldown = 100
        self._prev_fast = fast
        self._prev_slow = slow

    def CreateClone(self):
        return macd_1_min_scalper_strategy()
